addon_imps/storage/owncloud.py

```diff
@@ -0,0 +1,227 @@
+            response_xml = await response.text_content()
+            current_user_principal_url = self._parse_current_user_principal(
+                response_xml
+            )
+
+        if current_user_principal_url.startswith("/"):
+            current_user_principal_url = current_user_principal_url.lstrip("/")
+
+        async with self.network.PROPFIND(
+            uri_path=current_user_principal_url,
+            headers=headers,
+            content=self._build_propfind_displayname(),
+        ) as response:
+            response_xml = await response.text_content()
+            displayname = self._parse_displayname(response_xml)
+            return displayname
+
+    async def list_root_items(self, page_cursor: str = "") -> storage.ItemSampleResult:
+        return await self.list_child_items(_owncloud_root_id(), page_cursor)
+
+    async def get_item_info(self, item_id: str) -> storage.ItemResult:
+        item_type, path = _parse_item_id(item_id)
+        url = self._build_url(path)
+
+        headers = {
+            "Depth": "0",
+        }
+
+        async with self.network.PROPFIND(
+            uri_path=url,
+            headers=headers,
+            content=self._build_propfind_allprops(),
+        ) as response:
+            response_xml = await response.text_content()
+            parsed_item = self._parse_item(response_xml, path)
+            return parsed_item
+
+    async def list_child_items(
+        self,
+        item_id: str,
+        page_cursor: str = "",
+        item_type: storage.ItemType | None = None,
+    ) -> storage.ItemSampleResult:
+        _item_type, path = _parse_item_id(item_id)
+        url = self._build_url(path)
+
+        headers = {
+            "Depth": "1",
+        }
+
+        async with self.network.PROPFIND(
+            uri_path=url,
+            headers=headers,
+            content=self._build_propfind_allprops(),
+        ) as response:
+            response_xml = await response.text_content()
+            parsed_items = self._parse_items(
+                response_xml, base_path=path, item_type=item_type
+            )
+            return storage.ItemSampleResult(items=list(parsed_items))
+
+    def _build_url(self, path: str) -> str:
+        return path.lstrip("/")
+
+    def _build_propfind_current_user_principal(self) -> str:
+        return """<?xml version="1.0" encoding="UTF-8"?>
+        <d:propfind xmlns:d="DAV:">
+            <d:prop>
+                <d:current-user-principal/>
+            </d:prop>
+        </d:propfind>"""
+
+    def _build_propfind_displayname(self) -> str:
+        return """<?xml version="1.0" encoding="UTF-8"?>
+        <d:propfind xmlns:d="DAV:">
+            <d:prop>
+                <d:displayname/>
+            </d:prop>
+        </d:propfind>"""
+
+    def _build_propfind_allprops(self) -> str:
+        return """<?xml version="1.0" encoding="UTF-8"?>
+        <d:propfind xmlns:d="DAV:">
+            <d:allprop/>
+        </d:propfind>"""
```

Comment:
Maybe extract this into Enum or at least keep them separate from Imp as they impose some clutter